spec/bulk_processor_spec.rb

```diff
@@ -49,97 +48,55 @@ class TestHandler < MockHandler
-
-        expect(processor.process).to eq(true)
-        expect(enqueued_jobs.length).to eq(1)
-      end
-
-      it 'strips non-UTF-8 characters from the stream' do
-        perform_enqueued_jobs do
-          stream = StringIO.new("name\nyen=\xA5")
-          expect(TestItemProcessor)
-            .to receive(:new).with({ 'name' => 'yen=' }, anything).and_call_original
-          processor = BulkProcessor.new(stream, TestItemProcessor, TestHandler)
-          processor.process
+          expect(subject.start).to eq(false)
+          expect(subject.errors).to include(message)
         end
       end
     end
 
     context 'sucessfully processed the file' do
-      it 'calls .complete on the handler with the original payload' do
-        perform_enqueued_jobs do
-          expect(TestHandler)
-            .to receive(:complete).with({ my: 'stuff' }, anything, anything, nil)
-          processor =
-            BulkProcessor.new(stream, TestItemProcessor, TestHandler, { my: 'stuff' })
-          processor.process
-        end
+      subject do
+        BulkProcessor.new(stream: stream, processor_class: MockCSVProcessor,
+                          payload: { my: 'stuff' })
       end
 
-      it 'calls .complete on the handler with successes' do
+      it 'calls #complete! on the handler' do
         perform_enqueued_jobs do
-          expect(TestHandler)
-            .to receive(:complete).with(anything, { 0 => [] }, anything, nil)
-          processor = BulkProcessor.new(stream, TestItemProcessor, TestHandler)
-          processor.process
-        end
-      end
-
-      context 'but one row fails' do
-        let(:csv) { "name\nSocks" }
-
-        it 'calls .complete on the handler with failures' do
-          perform_enqueued_jobs do
-            expect(TestHandler)
-              .to receive(:complete)
-                    .with(anything, anything, { 0 => ['bad dog'] }, nil)
-            processor = BulkProcessor.new(stream, TestItemProcessor, TestHandler)
-            processor.process
-          end
+          expect(handler).to receive(:complete!)
+          subject.start
         end
       end
     end
 
     context 'failed to process the whole file' do
-      let(:csv) { "name\nHuman" }
+      before do
+        allow_any_instance_of(MockRowProcessor)
+          .to receive(:process!).and_raise(StandardError, 'Uh oh!')
+      end
 
-      it 'calls .complete with the fatal error' do
+      it 'calls #fail! with the fatal error' do
         perform_enqueued_jobs do
-          expect(TestHandler)
-            .to receive(:complete)
-                  .with(anything, anything, anything, instance_of(RuntimeError))
-          processor = BulkProcessor.new(stream, TestItemProcessor, TestHandler)
-          processor.process
+          begin
+            expect(handler).to receive(:fail!).with(instance_of(StandardError))
+            subject.start
+          rescue SignalException
```

Comment:
In this case, nothing should be raised, should it? Not a `SignalException` in any case.